### DISCUSSION TITLE
Sources - alternate impl

### DISCUSSION THREAD

**Discussion by BryanJacobs** [martinpaljak/FIDO2#4] [Show and tell]
Hey, it's been around six months with no sources here, so I wrote a thing that does this:

https://github.com/BryanJacobs/FIDO2Applet

I'm not trying to steal your thunder, and I still hope you release this better-sounding implementation, but I really wanted to have my smartcards working in the modern world. So I wrote some spaghetti.

Maybe parts of it would be useful for you?

**Comment by martinpaljak** [comment]
Cool! I think the design goals are a bit different, but it would be interesting if you wrapped your applet/emulator into a simple TCP server with a simple hex line protocol. I can put the minimalscaffolding for jcardsim (single file) into a gist if that would help.

I usually do a self-contained executable for an applet that brings up simple server (TCP, websocket, depends on context) that hosts an emulated card. This was the applet could be tested/introspected with the help of currently available fido utility with the `-T` option.

**Comment by BryanJacobs** [comment]
It can already do that just fine using `jCardSim` and there are docs in the repository for how to test it that way. In fact, I did most of the development testing with "real" FIDO2 tools. At first the CLI ones, but now it's at the point where I test with end-user applications like OpenSSH.

I don't build a "self-contained executable" but I'm not sure what benefit that would offer over a .cap and a java runscript. Unless there are no software FIDO2 tokens out there? Surely there must be one.

Thanks for `gpp`, by the way. Definitely made my life easier.

I hope that if you decide to finish your own implementation, mine is of some use to you. The FIDO2 standard has a moderate number of rough edges I had to work through.
